core/build.gradle.kts

```diff
@@ -187,6 +194,65 @@ ide {
     generatedSource(javaCCTest, "test")
 }
 
+fun configureAnnotationSet(task: JavaCompile, sourceSet: SourceSet) {
+    with(task) {
+        source = sourceSet.java
+        classpath = sourceSet.compileClasspath
+        destinationDirectory.set(options.generatedSourceOutputDirectory)
+        options.compilerArgs.add("-proc:only")
+        org.gradle.api.plugins.internal.JvmPluginsHelper.configureAnnotationProcessorPath(sourceSet, sourceSet.java, options, project)
+    }
+}
+
+val annotationProcessorMain by tasks.registering(JavaCompile::class) {
+    dependsOn(javaCCMain)
+    configureAnnotationSet(this, sourceSets.main.get())
+    // only if we aren't running compileJava, since doing twice fails.
+    onlyIf { tasks.findByPath("compileJava")?.enabled != true }
+}
+
+val annotationProcessorTest by tasks.registering(JavaCompile::class) {
+    val kotlinTestCompile = tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>()
+        .getByName("compileTestKotlin")
+
+    dependsOn(javaCCTest, kotlinTestCompile)
```

Comment:
Could you please clarify why compileTestKotlin is required?
It is sad because it would require "compile linq4j" and so on, so it would probably take noticeable time

Reply:
The annotation processor fails because there are references to kotlin based annotations in the test. It's possible that a change to classpath would resolve since I don't think the Kotlin annotations are Calcite specific (correct me if I'm wrong). If they are general Kotlin, how do I get the right classpath to add?

Reply:
Ah, it comes from https://github.com/apache/calcite/tree/master/core/src/test/kotlin/org/apache/calcite/testlib

The way out is probably to call annotation processing via [kapt](https://kotlinlang.org/docs/kapt.html#top). It looks like aptMode=stubs or aptMode=apt might be helpful.

----

Just in case, `:core:compileTestKotlin` requires `:core:classes`, so if you are going to trigger `compileTestKotlin` on IDE sync, then the dance with `-proc:only` is not really needed.